### PR TITLE
add jittered backoff for elb

### LINKED ISSUE #51108: Add AWSRetry to AWS modules where it is still missing
##### SUMMARY
The `AWSRetry` module provides the `backoff`, `exponential_backoff` and `jitter_backoff` methods to handle AWS API Throttling errors by retrying the call repeatedly with pauses.

Some AWS cloud modules have been adapted and changed to include the use of `AWSRetry`, while others are still missing this module.

List of modules that are yet to use `AWSRetry`:

```
lib/ansible/modules/cloud/amazon/aws_application_scaling_policy.py
lib/ansible/modules/cloud/amazon/aws_az_facts.py
lib/ansible/modules/cloud/amazon/aws_batch_compute_environment.py
lib/ansible/modules/cloud/amazon/aws_batch_job_definition.py
lib/ansible/modules/cloud/amazon/aws_batch_job_queue.py
lib/ansible/modules/cloud/amazon/aws_caller_facts.py
lib/ansible/modules/cloud/amazon/aws_codecommit.py
lib/ansible/modules/cloud/amazon/aws_direct_connect_gateway.py
lib/ansible/modules/cloud/amazon/aws_eks_cluster.py
lib/ansible/modules/cloud/amazon/aws_elasticbeanstalk_app.py
lib/ansible/modules/cloud/amazon/aws_glue_connection.py
lib/ansible/modules/cloud/amazon/aws_glue_job.py
lib/ansible/modules/cloud/amazon/aws_kms.py
lib/ansible/modules/cloud/amazon/aws_region_facts.py
lib/ansible/modules/cloud/amazon/aws_s3.py
lib/ansible/modules/cloud/amazon/aws_s3_bucket_facts.py
lib/ansible/modules/cloud/amazon/aws_s3_cors.py
lib/ansible/modules/cloud/amazon/aws_sgw_facts.py
lib/ansible/modules/cloud/amazon/aws_ssm_parameter_store.py
lib/ansible/modules/cloud/amazon/aws_waf_facts.py
lib/ansible/modules/cloud/amazon/aws_waf_rule.py
lib/ansible/modules/cloud/amazon/aws_waf_web_acl.py
lib/ansible/modules/cloud/amazon/cloudfront_distribution.py
lib/ansible/modules/cloud/amazon/cloudfront_facts.py
lib/ansible/modules/cloud/amazon/cloudfront_invalidation.py
lib/ansible/modules/cloud/amazon/cloudfront_origin_access_identity.py
lib/ansible/modules/cloud/amazon/cloudtrail.py
lib/ansible/modules/cloud/amazon/cloudwatchevent_rule.py
lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py
lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group_facts.py
lib/ansible/modules/cloud/amazon/data_pipeline.py
lib/ansible/modules/cloud/amazon/dynamodb_table.py
lib/ansible/modules/cloud/amazon/dynamodb_ttl.py
lib/ansible/modules/cloud/amazon/ec2.py
lib/ansible/modules/cloud/amazon/ec2_ami.py
lib/ansible/modules/cloud/amazon/ec2_ami_copy.py
lib/ansible/modules/cloud/amazon/ec2_ami_facts.py
lib/ansible/modules/cloud/amazon/ec2_asg_facts.py
lib/ansible/modules/cloud/amazon/ec2_asg_lifecycle_hook.py
lib/ansible/modules/cloud/amazon/ec2_customer_gateway_facts.py
lib/ansible/modules/cloud/amazon/ec2_eip.py
lib/ansible/modules/cloud/amazon/ec2_eip_facts.py
lib/ansible/modules/cloud/amazon/ec2_elb.py
lib/ansible/modules/cloud/amazon/ec2_elb_lb.py
lib/ansible/modules/cloud/amazon/ec2_eni.py
lib/ansible/modules/cloud/amazon/ec2_eni_facts.py
lib/ansible/modules/cloud/amazon/ec2_group_facts.py
lib/ansible/modules/cloud/amazon/ec2_instance_facts.py
lib/ansible/modules/cloud/amazon/ec2_key.py
lib/ansible/modules/cloud/amazon/ec2_lc.py
lib/ansible/modules/cloud/amazon/ec2_lc_facts.py
lib/ansible/modules/cloud/amazon/ec2_lc_find.py
lib/ansible/modules/cloud/amazon/ec2_metadata_facts.py
lib/ansible/modules/cloud/amazon/ec2_metric_alarm.py
lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py
lib/ansible/modules/cloud/amazon/ec2_scaling_policy.py
lib/ansible/modules/cloud/amazon/ec2_snapshot.py
lib/ansible/modules/cloud/amazon/ec2_snapshot_copy.py
lib/ansible/modules/cloud/amazon/ec2_snapshot_facts.py
lib/ansible/modules/cloud/amazon/ec2_tag.py
lib/ansible/modules/cloud/amazon/ec2_vol.py
lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_dhcp_option.py
lib/ansible/modules/cloud/amazon/ec2_vpc_dhcp_option_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
lib/ansible/modules/cloud/amazon/ec2_vpc_endpoint.py
lib/ansible/modules/cloud/amazon/ec2_vpc_igw_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_nacl.py
lib/ansible/modules/cloud/amazon/ec2_vpc_nacl_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway.py
lib/ansible/modules/cloud/amazon/ec2_vpc_nat_gateway_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_peer.py
lib/ansible/modules/cloud/amazon/ec2_vpc_peering_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_vgw_facts.py
lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py
lib/ansible/modules/cloud/amazon/ec2_vpc_vpn_facts.py
lib/ansible/modules/cloud/amazon/ec2_win_password.py
lib/ansible/modules/cloud/amazon/ecs_attribute.py
lib/ansible/modules/cloud/amazon/ecs_cluster.py
lib/ansible/modules/cloud/amazon/ecs_ecr.py
lib/ansible/modules/cloud/amazon/ecs_service.py
lib/ansible/modules/cloud/amazon/ecs_task.py
lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
lib/ansible/modules/cloud/amazon/ecs_taskdefinition_facts.py
lib/ansible/modules/cloud/amazon/efs.py
lib/ansible/modules/cloud/amazon/elasticache.py
lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py
lib/ansible/modules/cloud/amazon/elasticache_snapshot.py
lib/ansible/modules/cloud/amazon/elasticache_subnet_group.py
lib/ansible/modules/cloud/amazon/elb_application_lb.py
lib/ansible/modules/cloud/amazon/elb_application_lb_facts.py
lib/ansible/modules/cloud/amazon/elb_classic_lb.py
lib/ansible/modules/cloud/amazon/elb_instance.py
lib/ansible/modules/cloud/amazon/elb_network_lb.py
lib/ansible/modules/cloud/amazon/elb_target.py
lib/ansible/modules/cloud/amazon/elb_target_group.py
lib/ansible/modules/cloud/amazon/elb_target_group_facts.py
lib/ansible/modules/cloud/amazon/execute_lambda.py
lib/ansible/modules/cloud/amazon/iam.py
lib/ansible/modules/cloud/amazon/iam_cert.py
lib/ansible/modules/cloud/amazon/iam_mfa_device_facts.py
lib/ansible/modules/cloud/amazon/iam_policy.py
lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py
lib/ansible/modules/cloud/amazon/iam_user.py
lib/ansible/modules/cloud/amazon/kinesis_stream.py
lib/ansible/modules/cloud/amazon/lambda.py
lib/ansible/modules/cloud/amazon/lambda_alias.py
lib/ansible/modules/cloud/amazon/lambda_event.py
lib/ansible/modules/cloud/amazon/lambda_facts.py
lib/ansible/modules/cloud/amazon/lambda_policy.py
lib/ansible/modules/cloud/amazon/lightsail.py
lib/ansible/modules/cloud/amazon/rds_param_group.py
lib/ansible/modules/cloud/amazon/rds_subnet_group.py
lib/ansible/modules/cloud/amazon/redshift.py
lib/ansible/modules/cloud/amazon/redshift_cross_region_snapshots.py
lib/ansible/modules/cloud/amazon/redshift_facts.py
lib/ansible/modules/cloud/amazon/redshift_subnet_group.py
lib/ansible/modules/cloud/amazon/route53.py
lib/ansible/modules/cloud/amazon/route53_facts.py
lib/ansible/modules/cloud/amazon/route53_health_check.py
lib/ansible/modules/cloud/amazon/route53_zone.py
lib/ansible/modules/cloud/amazon/s3_lifecycle.py
lib/ansible/modules/cloud/amazon/s3_logging.py
lib/ansible/modules/cloud/amazon/s3_sync.py
lib/ansible/modules/cloud/amazon/s3_website.py
lib/ansible/modules/cloud/amazon/sns.py
lib/ansible/modules/cloud/amazon/sqs_queue.py
lib/ansible/modules/cloud/amazon/sts_assume_role.py
lib/ansible/modules/cloud/amazon/sts_session_token.py
```

We have been experiencing the `RequestLimitExceeded` error from AWS while using Ansible very frequently recently, mostly coming from the various `*_facts` modules like `ec2_ami_facts`, `ec2_eip_facts`, and others. We also use `ec2_tag`, `ec2_lc`, `ec2_key`, `ec2_eip` frequently none of which are using the `AWSRetry` module.

##### ISSUE TYPE
- Asking to solve a very annoying problem

##### COMPONENT NAME
AWS Cloud Modules

### PR DESCRIPTION
**Pull request by fawaf**
##### SUMMARY

related to #51108

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

elb_target

##### ADDITIONAL INFORMATION

```
2020-01-21 10:40:55,705 p=29993 u=felixwong |  ok: [host14286 -> localhost] => (item=elb14286)
2020-01-21 10:40:55,718 p=29993 u=felixwong |  ok: [host14286 -> localhost] => (item=elb14286)
2020-01-21 10:40:55,738 p=29993 u=felixwong |  ok: [host14286 -> localhost] => (item=elb14286)
2020-01-21 10:40:55,744 p=29993 u=felixwong |  ok: [host14286 -> localhost] => (item=elb14286)
2020-01-21 10:40:55,750 p=29993 u=felixwong |  ok: [host14286 -> localhost] => (item=elb14286)
2020-01-21 10:40:55,757 p=29993 u=felixwong |  PLAY [HARD Router AMI Upgrade] *************************************************
2020-01-21 10:40:57,318 p=29993 u=felixwong |  TASK [Pre-deploy Apache httpd version] *****************************************
2020-01-21 10:40:57,321 p=29993 u=felixwong 
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/elb_target.py [hunk 1]
```diff
@@ -166,6 +167,7 @@ def describe_targets(connection, module, tg_arn, target):
                          exception=traceback.format_exc())
 
 
+@AWSRetry.jittered_backoff(retries=5, delay=5)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_target.py]: ClientError is always caught and handled in this function (and the one below) so these decorators have no effect. AWSRetry is intended to be used before ClientError is handled. You should wrap the individual API calls that are problematic inside of the exception handling.

A lot of AWS modules do this by creating a smaller function to retry and call inside of the exception handling:
```python
@AWSRetry.jittered_backoff(retries=5, delay=5)
def register_targets_with_backoff(connection, target, target_group_arn):
    connection.register_targets(TargetGroupArn=target_group_arn, Targets=[target])

def register_target(connection, module):
    ...
            try:
                register_targets_with_backoff(connection, target, target_group_arn)
                ...
            except ClientError as e:
    ...
```
Alternatively you could use the decorator inside of the exception handling:
```python
def register_target(connection, module):
    ...
            try:
                AWSRetry.jittered_backoff()(connection.register_targets)(TargetGroupArn=target_group_arn, Targets=[target])
                ...
            except ClientError as e:
    ...
```

There are additional details and examples here https://docs.ansible.com/ansible/latest/dev_guide/platforms/aws_guidelines.html#api-throttling-rate-limiting-and-pagination.

Comment (fawaf) [reply lib/ansible/modules/cloud/amazon/elb_target.py]: fixed.

**Comment by gnowxilef** [comment]
good catch. thanks @s-hertel

**Review by s-hertel — APPROVED** [review]
From a quick code review, this looks good now. It could also use a [changelog](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

**Comment by tremble** [comment]
As mentioned by @s-hertel I think the only thing missing is a changelog fragment.

https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs

**Comment by fawaf** [comment]
ah, do i do that? i'll add that.

**Comment by s-hertel** [comment]
@fawaf yes, please

**Comment by fawaf** [comment]
@s-hertel @tremble added changelog.

**Review by jillr — APPROVED** [review]
LGTM - thanks @fawaf

**Comment by tremble** [comment]
LGTM
Thanks for your time @fawaf

**Comment by fawaf** [comment]
rebased my pr, not sure if it still needs another review.

**Review by tremble — APPROVED** [review]
LGTM

Yes, when you rebase it needs re-approving.  If we're lucky ansibot should merge this for us.

**Pull request merged**